filament/src/PostProcessManager.cpp

```diff
@@ -40,12 +40,57 @@ using namespace math;
 using namespace backend;
 using namespace filament::details;
 
-void PostProcessManager::init(FEngine& engine) noexcept {
-    mEngine = &engine;
+// ------------------------------------------------------------------------------------------------
+
+PostProcessManager::PostProcessMaterial::PostProcessMaterial(FEngine& engine,
+        uint8_t const* data, size_t size) noexcept {
+    mMaterial = upcast(Material::Builder().package(data, size).build(engine));
+    mMaterialInstance = mMaterial->getDefaultInstance();
+    mProgram = mMaterial->getProgram(0);
+}
+
+PostProcessManager::PostProcessMaterial::PostProcessMaterial(
+        PostProcessManager::PostProcessMaterial&& rhs) noexcept {
+    using namespace std;
+    swap(mMaterial, rhs.mMaterial);
+    swap(mMaterialInstance, rhs.mMaterialInstance);
+    swap(mProgram, rhs.mProgram);
+}
+
+PostProcessManager::PostProcessMaterial& PostProcessManager::PostProcessMaterial::operator=(
+        PostProcessManager::PostProcessMaterial&& rhs) noexcept {
+    using namespace std;
+    swap(mMaterial, rhs.mMaterial);
+    swap(mMaterialInstance, rhs.mMaterialInstance);
+    swap(mProgram, rhs.mProgram);
+    return *this;
+}
+
+PostProcessManager::PostProcessMaterial::~PostProcessMaterial() {
+    assert(mMaterial == nullptr);
+}
+
+void PostProcessManager::PostProcessMaterial::terminate(FEngine& engine) noexcept {
+    engine.destroy(mMaterial);
+    mMaterial = nullptr;
+    mMaterialInstance = nullptr;
+    mProgram.clear();
+}
+
+// ------------------------------------------------------------------------------------------------
+
+PostProcessManager::PostProcessManager(FEngine& engine) noexcept : mEngine(engine) {
+}
+
+void PostProcessManager::init() noexcept {
     mPostProcessUb = UniformBuffer(PostProcessingUib::getUib().getSize());
 
+    mSSAO = PostProcessMaterial(mEngine, MATERIALS_SAO_DATA, MATERIALS_SAO_SIZE);
```

Comment:
We should measure at some point the cost of loading all the built-in materials, it will increase initialization time.

Reply:
Good point